src/warc2zim/main.py

```diff
@@ -219,64 +259,201 @@ def add_remote_or_local(self, filename):
         ) as zimcreator:
-            # add replay system
-            for article in self.replay_articles:
-                zimcreator.add_article(article)
-
-            for warcfile in self.inputs:
-                self.process_warc(warcfile, zimcreator)
-
-            # process revisits, headers only
-            for url, record in self.revisits.items():
-                if url not in self.indexed_urls:
-                    logger.debug(
-                        "Adding revisit {0} -> {1}".format(
-                            url, record.rec_headers["WARC-Refers-To-Target-URI"]
-                        )
-                    )
-                    zimcreator.add_article(WARCHeadersArticle(record))
+            # zimcreator.update_metadata(**self.metadata)
+
+            for article in self.generate_all_articles():
+                if article:
+                    zimcreator.add_zim_article(article)
 
-    def process_warc(self, warcfile, zimcreator):
-        with open(warcfile, "rb") as warc_fh:
-            for record in ArchiveIterator(warc_fh):
-                try:
+    def iter_warc_records(self):
+        for warcfile in self.inputs:
+            with open(warcfile, "rb") as warc_fh:
+                for record in ArchiveIterator(warc_fh):
                     if record.rec_type not in ("resource", "response", "revisit"):
                         continue
 
-                    url = record.rec_headers["WARC-Target-URI"]
-                    if url in self.indexed_urls:
-                        logger.warning(
-                            "Skipping duplicate {0}, already added to ZIM".format(url)
-                        )
-                        continue
+                    yield record
+
+    def find_main_page_metadata(self):
+        for record in self.iter_warc_records():
+            if record.rec_type == "revisit":
+                continue
+
+            # if no main_url, use first 'text/html' record as the main page by default
+            # not guaranteed to always work
+            mime = get_record_mime_type(record)
+
+            url = record.rec_headers["WARC-Target-URI"]
+
+            if (
+                not self.main_url
+                and mime == "text/html"
+                and record.payload_length != 0
+                and (
+                    not record.http_headers
+                    or record.http_headers.get_statuscode() == "200"
+                )
+            ):
+                self.main_url = url
+                if not self.include_all and not self.include_domains:
+                    self.include_domains = [urlsplit(self.main_url).netloc]
+
+            if self.main_url != url:
+                continue
+
+            if mime not in HTML_TYPES:
+                logger.warning(
+                    "Main page is not an HTML Page, mime type is: {0} - Skipping Favicon and Language detection".format(
+                        mime
+                    )
+                )
+
```

Comment:
didn't you miss a `continue` here?

Reply:
Added some clarification.. The warning is to indicate the main page url is not an html page, it'll still be allowed (i guess could be image or pdf or something) but is not technically an error..